design/tg-branches.md

```diff
@@ -0,0 +1,111 @@
+# TestGrid Branch Support 
+
+As automated PRs come in for Kurl features, such as add-on upgrades, there also needs to be testing automation in place to have confidence in the changes without manual testing. 
+This proposal suggest how to extend TestGrid to support execution against branches and github actions that will initiate `tgrun` against the PR branch.
+
+
+## Goals
+- Enable TestGrid to work against any branch.
+- Provide good test coverage of changes with as little unnecessary testing as possible.
+
+## Non Goals
+
+- Automating kurl addon create or upgrades (see [PR-1014](https://github.com/replicatedhq/kURL/pull/1014) )
+
+## Background
+
+As of today TestGrid runs nightly (prod), on merge to master (staging) and on-demand using `tgrun`.
+All this happens against the master branch and runs through the productions testgrid.kurl.sh server.
+The `tgrun` cli uses kurl.sh to generate a unique kurl installer for each run instance, and run instances use kurl.sh and the distribution packages in S3 to execute the test.   
+
+![TestGrid Diagram](./assets/Testgrid%20Nighty.png)
+
+Scenarios to address:
+1. Automated addon update
+1. Testing a new addon
+1. K8S update (automated?)
+
+Problems that needed to be addressed in the design:
+1. Where will untested/alpha packages live, and how do they get there?
+1. Which environment(s) do PR TestGrid runs execute against?
+1. How to get a Kurl installer for packages that aren't yet supported (i.e. break validation)?
+1. How navigate addons being hard-coded in tools and scripts?
+
+## High-Level Design
+
+Use the following environments:
+1. Use Prod TestGrid for executing PR tests
+1. Use staging.kurl.sh for installers
+
+Code changes that are required:
+1. The Kurl installer spec will be updated to allow installers to be generated with `upsteam_override`.
+1. kurl.sh will be modified to allow any version for an addon in the spec with the `upstream_override` enabled.
+1. `tgrun` will be updated
+   1. Since this proposal will only really test updated addons, a flag to run only a single "latest" kurl installer against all OSes.
+   1. Flags to specify addon version and download url
+1. Modify install.sh to use `upstream_override` bucket path from the kurl spec.
+   1. Possibly modify `yamltobash` utils (and others?)
+1. Add a github action to run on a PR creation/update.
+   1. Do a diff and see if any of the addon dirs have been modified.
+   1. Push packages to `<s3>/pr` for modified dirs (wait?)
+   1. `tgrun queue --ref=<short commit>-pr-<#> --latest-only --version=contour=1.12.0 --upstream_override=<s3 url in previous step>`
```

Comment:
How does tgrun know which addons to override if there's a single flag `--upstram_override`. Wouldn't you need an upstream override flag for each addon?

Reply:
Could you make tgrun accept a spec?

Reply:
@areed good look. I was thinking it would look like the `version` flag: `--upstream_override=contour=<s3 url>`
Maybe better names would be `version_override` and `s3_override`?

@emosbaugh I think thank could work. Like a `kubectl patch`, right? I think the only thing I don't like about that is then I think you need a spec file hanging the repo around for convenience, but maybe there is a better.